### PR TITLE
fix: add support for sha512-only archives

### PR DESCRIPTION
**Pull request by lczyk**
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

Attempted fix for https://github.com/canonical/chisel/issues/305.

Resolve digests through a small preference list (strongest first: `SHA512`, then `SHA256`) instead of hardcoding one field, and carry the hash kind alongside the digest so downloads are verified and cached with the matching algorithm. This covers:

- the Release index table (`Packages` digest),
- the per-package digest (the `.deb`), and
- the `Acquire-By-Hash` path merged from main, which now builds `by-hash/<algo>/<digest>` URLs from the resolved field rather than assuming `by-hash/SHA256/`.

Archives only guarantee a by-hash directory for the *strongest* hash they advertise, so the by-hash URL must be built from that one -- thanks @FinnRG for catching this. Since the preference list is strongest-first, the resolved field is already the right one, and a single ordering serves both purposes.

As of this writing Stonking publishes SHA512 only, in the base suite and in `-security`/`-updates` alike, so the whole suite resolves via SHA512. Resolute and earlier remain SHA256-only

An archive advertising *both* digests, content is now verified and cached under its SHA512 digest rather than its SHA256 one. Cache entries written by an older chisel are keyed by SHA256, so they go cold and are fetched once more, then expire normally (which, at the moment means they don't expire https://github.com/canonical/chisel/issues/261, but that's a separate issue). Nothing gets corrupted. Importantly, at present no archive Chisel targets advertises both digests (Resolute is `SHA256`-only, Stonking is `SHA512`-only), so in practice this is not a behaviour change.

### Testing

Added unit tests for the new behaviour and tested manually against the Stonking archive. The by-hash tests cover the four archive shapes: a by-hash directory for the strongest hash only, directories for several hashes (as Stonking actually publishes), a directory for a weaker hash only (a buggy archive, falls back to the named path), and no by-hash directories at all.

### Out of scope

Something i've deliberately decided is out of scope but really **should be fixed** in a followup: The manifest's per-package digest still reads `SHA256`, so on a SHA512-only archive it still errors. Most `chisel cut`'s still work unless the user requests the manifest to be cut too.

### CODE REVIEW

File: internal/archive/archive.go [hunk 1]
```diff
@@ -328,10 +331,46 @@ func (index *ubuntuIndex) fetchRelease() error {
...
+// preference. SHA256 is first so existing archives keep their cache keys;
+// Ubuntu 26.10 and later publish SHA512-only indices, handled by the fallback.
+var digestFields = []digestField{
+	{"SHA256", cache.SHA256},
+	{"SHA512", cache.SHA512},
+}
+
+// releaseDigest returns the checksum recorded for path in a Release "<hash>
+// <size> <path>" table, along with the field it was found in.
+func releaseDigest(release control.Section, path string) (digest string, field digestField) {
```
Comment (upils) [inline internal/archive/archive.go]: To be consistent with the rest of the code I think the "field" term should be replaced with "section".

File: internal/archive/testarchive/testarchive.go [hunk 2]
```diff
@@ -63,6 +64,8 @@ type Package struct {
 	Arch      string
 	Component string
 	Data      []byte
+	// Digest names the checksum field published for this package ("SHA256" when empty).
+	Digest string
```
Comment (lczyk) [inline internal/archive/testarchive/testarchive.go]: arguably this could be `Hashes []string` with some design tweaks

File: internal/archive/testarchive/testarchive.go [hunk 3]
```diff
@@ -107,6 +110,8 @@ type Release struct {
 	Label   string
 	Items   []Item
 	PrivKey *packet.PrivateKey
+	// Digest names the checksum field published for the index table ("SHA256" when empty).
+	Digest string
```
Comment (upils) [inline internal/archive/testarchive/testarchive.go]: Out of consistency this is not the digest but the digest kind. Same comment for the new field in `Package` above.

File: internal/archive/testarchive/testarchive.go [hunk 4]
```diff
@@ -216,6 +221,22 @@ func makeSha256(b []byte) string {
...
+		return "SHA256"
+	}
+	return kind
+}
+
+func makeDigest(kind string, b []byte) string {
+	if kind == "SHA512" {
+		return fmt.Sprintf("%x", sha512.Sum512(b))
+	}
+	return makeSha256(b)
```
Comment (upils) [inline internal/archive/testarchive/testarchive.go]: This is now the only call site of `makeSha256`, so it could be inlined here.

File: internal/archive/archive_test.go [hunk 5]
```diff
@@ -158,6 +158,16 @@ func (s *httpSuite) prepareArchiveAdjustRelease(suite, version, arch string, com
...
+	// Packages inherit the release's digest kind unless they set their own.
+	err = release.Walk(func(item testarchive.Item) error {
+		if p, ok := item.(*testarchive.Package); ok && p.Digest == "" {
+			p.Digest = release.Digest
+		}
+		return nil
+	})
+	if err != nil {
+		panic(err)
+	}
```
Comment (upils) [inline internal/archive/archive_test.go]: It looks a bit risky to do that *after* calling `adjustRelease` as the intention of the provided `adjustRelease` could be to tweak the packages digest. And so this additional piece of code could effectively undo it if the intention is to test a case with an empty Digest value. I recognize that is it a bit of a chicken/egg problem though, because in the new tests the code is actually relying on the result of `adjustRelease` to "propagate" the digest kind value. 

It works now but could lead to tricky bugs.

File: internal/archive/testarchive/testarchive.go (hunk 2 above)
Comment (upils) [reply internal/archive/testarchive/testarchive.go]: I agree, because otherwise it means we never test the case where the release actually publishes both digests.

**Review by upils — CHANGES_REQUESTED** [review]
Thanks @lczyk! This already looks very good.

I agree that, as you pointed out in the PR description, a follow-up PR must be created to deal with the manifest generation to consider this bug completely fixed. I can do it once this PR is merged, as we need to make sure it will work nicely with the current work on supporting bins.

File: internal/archive/archive.go (hunk 1 above)
Comment (lczyk) [reply internal/archive/archive.go]: https://github.com/canonical/chisel/pull/306/changes/1e1a0d14eeb7767bffbaf3e8a2ceddf4d48be183

File: internal/archive/testarchive/testarchive.go (hunk 2 above)
Comment (lczyk) [reply internal/archive/testarchive/testarchive.go]: https://github.com/canonical/chisel/pull/306/commits/46fefb3b2cbc83986e195044a5e549ec62101a8e

File: internal/archive/testarchive/testarchive.go (hunk 4 above)
Comment (lczyk) [reply internal/archive/testarchive/testarchive.go]: https://github.com/canonical/chisel/pull/306/commits/58c0e97fb7c348d0533be6cf8b19c7e61dea0be9

File: internal/archive/archive_test.go (hunk 5 above)
Comment (lczyk) [reply internal/archive/archive_test.go]: yeah. fair. i've reworked this bit a bit. digest kinds are now per-archive (not per-package) and get copied into the package at render time so can't be clobbered. also dropped the implicit empty -> SHA256 default. instead its explicitly set by `prepareArchiveAdjustRelease`.

**Comment by lczyk** [comment]
@upils ok, ready for another review round. `+302 -29` is on the bigger side but vast majority is tests, so i hope this is still ok

File: internal/archive/archive.go [hunk 6]
```diff
@@ -328,10 +331,46 @@ func (index *ubuntuIndex) fetchRelease() error {
...
+	// No digest advertised; fall back to SHA256 so the package can still be
+	// cached and retrieved by its computed digest.
+	return "", cache.SHA256
+}
+
 func (index *ubuntuIndex) fetchIndex() error {
-	releaseDigests := index.release.Get("SHA256")
 	packagesPath := fmt.Sprintf("%s/binary-%s/Packages", index.component, index.arch)
-	packagesDigest, _, _ := control.ParsePathInfo(releaseDigests, packagesPath)
+	packagesDigest, section := releaseDigest(index.release, packagesPath)
```
Comment (upils) [inline internal/archive/archive.go]: Sorry, I was wrong to suggest renaming it to section. The name is now ambiguous as the term is used to name other things (actual control.Section) in the package. I now think you were in the right track with the "field" term.

File: internal/archive/testarchive/testarchive.go [hunk 7]
```diff
@@ -212,7 +249,10 @@ func (pi *PackageIndex) Content() []byte {
 	return MergeSections(pi.Packages)
 }
 
-func makeSha256(b []byte) string {
+func makeDigest(kind string, b []byte) string {
```
Comment (upils) [inline internal/archive/testarchive/testarchive.go:251]: Now this function has an implicit default. If it is called with `makeDigest("MD5", []byte("something"))` it will return a sha256 sum, which does not make a lot of sense. Maybe it can be converted to a `switch` statement and, as with some other helpers in this package, panic when the digest is unknown?

File: internal/archive/testarchive/testarchive.go [hunk 8]
```diff
@@ -124,11 +140,30 @@ func (r *Release) Section() []byte {
...
+			if p, ok := item.(*Package); ok {
+				p.digestKinds = r.DigestKinds
+			}
+			return nil
+		})
+	}
+}
+
 func (r *Release) Content() []byte {
+	r.inheritDigestKinds()
```
Comment (upils) [inline internal/archive/testarchive/testarchive.go]: This call seems superfluous. Only calling `inheritDigestKinds()` in `Render` is enough. I tested it to be sure and tests pass. What that intentional to cover a corner case I am missing though?

File: internal/archive/testarchive/testarchive.go [hunk 9]
```diff
@@ -124,11 +140,30 @@ func (r *Release) Section() []byte {
 	return nil
 }
 
+// inheritDigestKinds copies the release's digest kinds onto every package it
+// publishes, so the archive-wide choice reaches each package section. Run at
+// render time and idempotent, so it always sees the final release state.
+func (r *Release) inheritDigestKinds() {
+	for _, item := range r.Items {
+		// The callback never errors, so neither can Walk.
```
Comment (upils) [inline internal/archive/testarchive/testarchive.go]: I am on the fence on this one. On one hand you are right today, and the implementation is short and easy to read. On the other hand, returning the error and panicking at the call sites seems safer and more consistent with the rest of the code. It would protect us the day the walk function is reworked and is actually returning an error.

File: internal/archive/testarchive/testarchive.go [hunk 10]
```diff
@@ -63,6 +64,10 @@ type Package struct {
 	Arch      string
 	Component string
 	Data      []byte
+	// digestKinds is filled from the release at render time (see
+	// inheritDigestKinds). Digest kinds are an archive-wide choice, so tests set
+	// them on the Release, not per package.
```
Comment (upils) [inline internal/archive/testarchive/testarchive.go]: This seems to partially document the function manipulating this field and not the field itself.

**Review by upils — CHANGES_REQUESTED** [review]
Thanks @lczyk. I have more comments, and sorry for the back-and-forth on the section/field naming (see my comment below).

File: internal/archive/archive.go (hunk 6 above)
Comment (lczyk) [reply internal/archive/archive.go]: np. reverted in https://github.com/canonical/chisel/pull/306/commits/744fae97fe61bf39bd0727a07e33b9f45cac7b13

File: internal/archive/testarchive/testarchive.go (hunk 7 above)
Comment (lczyk) [reply internal/archive/testarchive/testarchive.go:251]: https://github.com/canonical/chisel/pull/306/commits/8c5cbc1e9ecc59b2e19303da0282f569c34f8b50

File: internal/archive/testarchive/testarchive.go (hunk 9 above)
Comment (lczyk) [reply internal/archive/testarchive/testarchive.go]: hmm.. yep, good point. refactored in https://github.com/canonical/chisel/pull/306/commits/2ec8ab9e8948df27c91d8f2236b7c7e37821b3c2. NOTE: i've threaded the err all the way out or `Render`.

File: internal/archive/testarchive/testarchive.go (hunk 8 above)
Comment (lczyk) [reply internal/archive/testarchive/testarchive.go]: it was to protect from ppl accidentally calling either of `Release.Content()`, `PackageIndex.Content()` or `Package.Section()` before the inheritance, but in the current code that does not happen. i've removed this in https://github.com/canonical/chisel/pull/306/commits/7a308a375bc6cfc07a127533289de0646e15f774 and the *tentatively* did https://github.com/canonical/chisel/pull/306/commits/13977093a0a336bd04f93f45ad34d4a3e8e42c27 but ff to (tell me to) revert it. not 100% sure about it

File: internal/archive/testarchive/testarchive.go (hunk 10 above)
Comment (lczyk) [reply internal/archive/testarchive/testarchive.go]: https://github.com/canonical/chisel/pull/306/commits/c5ce9e3ff17a92697016be33130cd4378fd08a39 but then redone in https://github.com/canonical/chisel/pull/306/commits/13977093a0a336bd04f93f45ad34d4a3e8e42c27

File: internal/archive/archive.go [hunk 11]
```diff
@@ -328,10 +331,57 @@ func (index *ubuntuIndex) fetchRelease() error {
...
+
+// digestFields lists the checksum fields Chisel can verify, strongest first.
+// This order matches the by-hash archive layout: an archive is only guaranteed
+// to publish a by-hash directory for the strongest hash it advertises.
+var digestFields = []digestField{
+	{"SHA512", cache.SHA512},
+	{"SHA256", cache.SHA256},
+}
+
+// cacheDigestFields is digestFields reordered to prefer SHA256, so that
```
Comment (upils) [inline internal/archive/archive.go]: cacheDigestFields is not the result of a function on digestFields so this comment is a bit misleading.

File: internal/archive/archive.go [hunk 12]
```diff
@@ -328,10 +331,57 @@ func (index *ubuntuIndex) fetchRelease() error {
...
+// doubles as the by-hash directory name in the archive layout.
+type digestField struct {
+	name string
+	kind cache.DigestKind
+}
+
+// digestFields lists the checksum fields Chisel can verify, strongest first.
+// This order matches the by-hash archive layout: an archive is only guaranteed
+// to publish a by-hash directory for the strongest hash it advertises.
+var digestFields = []digestField{
```
Comment (upils) [inline internal/archive/archive.go:345]: What do you think of renaming:
- digestFields -> byhashOrder
- cacheDigestFields -> cacheOrder

?

File: internal/archive/archive.go [hunk 13]
```diff
@@ -328,10 +331,57 @@ func (index *ubuntuIndex) fetchRelease() error {
...
+// TODO Drop this in favour of digestFields. That changes cache keys for
+// archives publishing both digests, so it must wait for a minor release.
+var cacheDigestFields = []digestField{
+	{"SHA256", cache.SHA256},
+	{"SHA512", cache.SHA512},
+}
+
+// findDigest returns the checksum recorded for path in a Release "<hash>
+// <size> <path>" table, along with the field it was found in, trying the
+// fields in order.
```
Comment (upils) [inline internal/archive/archive.go]: This mentions details abstracted by the usage of `control.ParsePathInfo`. The shape of the table in the Release file is of no concern of this function.

File: internal/archive/testarchive/testarchive.go [hunk 14]
```diff
@@ -124,11 +144,26 @@ func (r *Release) Section() []byte {
 	return nil
 }
 
+// adoptPackages points every published package back at this release, giving
+// the package sections access to the archive-wide digest kinds. Render calls
+// it before walking the items.
+func (r *Release) adoptPackages() error {
```
Comment (upils) [inline internal/archive/testarchive/testarchive.go]: When reading this signature, I think "this method is used for the release to adopt packages". In a sense this is true but the direction is Package -> Release. What about `wirePackages`, or anything that would describe the back-reference being established?

File: internal/archive/testarchive/testarchive.go [hunk 15]
```diff
@@ -63,6 +65,10 @@ type Package struct {
 	Arch      string
 	Component string
 	Data      []byte
+	// release is set when a release adopts this package at render time.
+	// Section reads the archive-wide digest kinds through it, so the package
+	// always sees the release's current choice.
+	release *Release
```
Comment (upils) [inline internal/archive/testarchive/testarchive.go]: I wonder why we now need to wire the whole release object here when we are only interested in the digest kinds. I saw your commit message but it is unclear why a copy at render time is not enough. Is the digest kinds slice changing over time in some cases?

File: internal/archive/testarchive/testarchive.go [hunk 16]
```diff
@@ -74,7 +80,17 @@ func (p *Package) Walk(f func(Item) error) error {
 }
 
 func (p *Package) Section() []byte {
+	if p.release == nil {
+		panic("package section rendered before a release adopted the package; render via Release.Render")
+	}
```
Comment (upils) [inline internal/archive/testarchive/testarchive.go]: `(p *Package) Section()` and `(r *Release) Render(...)` are now implicitly tied together and this check tries to prevent a misuse. Given our current usage it is fine but I wonder if that should be made more obvious somehow.

File: internal/archive/testarchive/testarchive.go [hunk 17]
```diff
@@ -175,8 +213,13 @@ func (r *Release) Render(prefix string, content map[string][]byte) error {
 		distItemPath := path.Join(prefix, "dists", r.Suite, itemPath)
 		content[distItemPath] = itemContent
 		if r.ByHash && itemPath != r.Path() {
-			byHashPath := path.Join(prefix, "dists", r.Suite, path.Dir(itemPath), "by-hash", "SHA256", makeSha256(itemContent))
-			content[byHashPath] = itemContent
+			// Real archives (ftpmaster) only publish by-hash directories for
+			// the strongest hash they advertise; mirror that so tests catch
+			// clients building by-hash URLs from a weaker hash.
```
Comment (upils) [inline internal/archive/testarchive/testarchive.go]: Real archives only guarantee by-hash directories for the strongest hash, but they can do it for weaker hashes too. Stonking is offering both SHA512 and SHA256 https://ports.ubuntu.com/ubuntu-ports/dists/stonking/by-hash/. I suppose we should be able to test that Chisel behaves correctly in all cases:
- by-hash directory for only the strongest
- by-hash directories for multiple hashes
- by-hash directory for one hash that is not the strongest one (a buggy archive then)
- no by-hash directories

Most of this is already tested but I don't think the second one is. Correct?

**Review by upils — CHANGES_REQUESTED** [review]
Thanks again @lczyk. This is indeed getting a bit bigger than we wanted but I do not see a sensible way to split this work further.

File: internal/archive/archive.go (hunk 12 above)
Comment (lczyk) [reply internal/archive/archive.go:345]: i have no strong feelings either way. i will rename it if you'd like, and i'll leave it otherwise. i guess at this point i git used to `digestFields` a bit, but that should not guide the design.

**Comment by lczyk** [comment]
> This is indeed getting a bit bigger than we wanted but I do not see a sensible way to split this work further.

yeah, alas neither. very open for ideas though, and trying to keep scope small and focused.

File: internal/archive/testarchive/testarchive.go (hunk 15 above)
Comment (lczyk) [reply internal/archive/testarchive/testarchive.go]: i thought a backpointer to Release was a cleaner way to implement the `if p.release == nil` guard.

with `DigestKinds []string`, the empty/nil slice is ambiguous: it means both _"archive publishes no digests"_ and _"not rendered yet"_, so we'd need another field (like `rendered bool`). with `release *Release` this is only one field ` since `p.release == nil` always means _"not rendered yet"_.

happy to change back to copy but then we grow one more field. if that's fine then no problem. 👍

File: internal/archive/testarchive/testarchive.go (hunk 17 above)
Comment (lczyk) [reply internal/archive/testarchive/testarchive.go]: i believe you're right

File: internal/archive/testarchive/testarchive.go (hunk 16 above)
Comment (lczyk) [reply internal/archive/testarchive/testarchive.go]: > wonder if that should be made more obvious somehow.

i've not found anything short of a more major refactor which threads digest kinds through Section and Content explicitly. i was trying to keep the PR small and still keep the guard, even if, for now, the guard is not structural but a panic. i thought this would be ok for now, given that the fail state of this is: _"the tests obviously don't work"_ rather than actually adding a panic to `chisel` binary itself

File: internal/archive/testarchive/testarchive.go (hunk 14 above)
Comment (lczyk) [reply internal/archive/testarchive/testarchive.go]: https://github.com/canonical/chisel/pull/306/commits/abfc91fdf60da4e5d30fce87210142f594e91c4d

File: internal/archive/archive.go (hunk 13 above)
Comment (lczyk) [reply internal/archive/archive.go]: yep. fair point.

https://github.com/canonical/chisel/pull/306/commits/0ebadf2d46ba40f347880ba53ba8a755853de086

File: internal/archive/archive.go (hunk 11 above)
Comment (lczyk) [reply internal/archive/archive.go]: hmm.. so, yes, but also i just had a second look at this and i think we could just use digestFields. the worst case is that we have some cache misses from old downloads from pre-this-version-(hopefully-1.4.3). if that's acceptable, it would simplify the PR a little bit.

Comment (lczyk) [reply internal/archive/archive.go]: also `cacheDigestFields` only really does anything when archive publishes both sha's and atm resolute is only sha256 (and md5 and sha1 which we dont care about) and stonking is only sha512

File: internal/archive/testarchive/testarchive.go (hunk 17 above)
Comment (lczyk) [reply internal/archive/testarchive/testarchive.go]: https://github.com/canonical/chisel/pull/306/commits/606e7a174b082b1fa9f3c4e1a800934c00afb05e

File: internal/archive/archive.go (hunk 11 above)
Comment (upils) [reply internal/archive/archive.go]: I think you are right. This is on paper a small change in behavior but it is not really in practice. I think it can be simplified to only use `digestFields`. Please make sure to mention that in the PR description.

File: internal/archive/archive.go (hunk 12 above)
Comment (upils) [reply internal/archive/archive.go:345]: Given my other comment below this point is now a non-issue if we end up with a single digestField slice.

File: internal/archive/testarchive/testarchive.go [hunk 18]
```diff
@@ -124,11 +144,26 @@ func (r *Release) Section() []byte {
 	return nil
 }
 
+// wirePackages points every published package back at this release, giving
+// the package sections access to the archive-wide digest kinds. Render calls
+// it before walking the items.
```
Comment (upils) [inline internal/archive/testarchive/testarchive.go]: This comments should probably not document the behavior of other functions using this function. 

What about:

```suggestion
// wirePackages gives the package sections access to the archive-wide digest kinds.
```

File: internal/archive/testarchive/testarchive.go (hunk 15 above)
Comment (upils) [reply internal/archive/testarchive/testarchive.go]: I agree with you, let's keep it.

File: internal/archive/testarchive/testarchive.go (hunk 16 above)
Comment (upils) [reply internal/archive/testarchive/testarchive.go]: I think that OK for now, I agree with you.

**Review by upils — CHANGES_REQUESTED** [review]
Thanks @lczyk. Let's do the simplification you proposed and this should be good for the core review.

File: internal/archive/archive.go (hunk 11 above)
Comment (lczyk) [reply internal/archive/archive.go]: https://github.com/canonical/chisel/pull/306/commits/311052ee500e1ec8929688cdfcd6cc5fd4749e4d

**Comment by lczyk** [comment]
@upils changes made and desc updated

**Review by upils — APPROVED** [review]
Thanks!

File: internal/archive/testarchive/testarchive.go (hunk 15 above)
Comment (niemeyer) [reply internal/archive/testarchive/testarchive.go]: This indeed does not feel so great. We are improving Chisel to generalize the digest logic where we assumed a constraint in the past, and on the way we are again establishing a new constraint that is not a ground truth for how archives are organized.

Just to make the point more clear, which release of Ubuntu is this:

https://www.archive.ubuntu.com/ubuntu/pool/main/l/linux/

Also, there is a question about the fact this is being done per release instead of per package. How did we transition between the hashes?  Was the entire release rebuilt at once, and has apt survived the transition phase without failures?  How?

File: internal/archive/testarchive/testarchive.go [hunk 19]
```diff
@@ -74,7 +80,17 @@ func (p *Package) Walk(f func(Item) error) error {
 }
 
 func (p *Package) Section() []byte {
+	if p.release == nil {
+		panic("package section rendered before a release adopted the package; render via Release.Render")
```
Comment (niemeyer) [inline internal/archive/testarchive/testarchive.go]: As another consequence, and I know this is clear by now, but just to state it: this used to pure a pure data type and with this change we cannot render a Package without a private field being set, and it panics instead of failing. If nothing else, it's more evidence that it seems to be making things less nice.

File: internal/archive/testarchive/testarchive.go [hunk 20]
```diff
@@ -175,8 +211,15 @@ func (r *Release) Render(prefix string, content map[string][]byte) error {
 		distItemPath := path.Join(prefix, "dists", r.Suite, itemPath)
 		content[distItemPath] = itemContent
 		if r.ByHash && itemPath != r.Path() {
-			byHashPath := path.Join(prefix, "dists", r.Suite, path.Dir(itemPath), "by-hash", "SHA256", makeSha256(itemContent))
-			content[byHashPath] = itemContent
+			// Archives only guarantee a by-hash directory for the strongest
+			// hash they advertise, though they may publish more. Render the
```
Comment (niemeyer) [inline internal/archive/testarchive/testarchive.go]: Is this algorithm how apt determines those paths as well?  Given the history of hash changes in several applications, I don't feel very safe to be making another guess that can easily break when we start to have multiple hash kinds for the same archives in the near future.

File: internal/archive/testarchive/testarchive.go [hunk 21]
```diff
@@ -175,8 +211,15 @@ func (r *Release) Render(prefix string, content map[string][]byte) error {
...
+			// guaranteed one only, so tests catch clients building by-hash
+			// URLs from a weaker hash; tests wanting extra directories add
+			// them on top of the rendered content.
+			if kind := strongestKind(r.DigestKinds); kind != "" {
```
Comment (niemeyer) [inline internal/archive/testarchive/testarchive.go]: It's at least worth of note that we are doing this for every individual element instead of the overall release. Is that how it works for the archives as well?  What happened during the transition between the the hash kinds, did we have half the release with one hash and half with the other?  Was apt making the right choices during that time?

I moved half of this question to earlier on in the review, so feel free to ignore the points already replied to.

**Review by niemeyer — CHANGES_REQUESTED** [review]
Thanks a lot for this. The code looks clean and simple, but some of the wiring choices might do with some improve.

Given this was already reviewed repeatedly, I suggest you pick up from here @upils.

File: internal/archive/testarchive/testarchive.go (hunk 15 above)
Comment (FinnRG) [reply internal/archive/testarchive/testarchive.go]: The ftpmaster publisher always rebuilds the entire Release (and associated index files) all at once, so there was no transition phase.

Just to be clear, the hashes inside the Packages and Sources index files may be different from the hash(es) used in the Release file and therefore the by-hash lookup.

File: internal/archive/archive_test.go [hunk 22]
```diff
@@ -129,10 +131,14 @@ func (s *httpSuite) prepareArchive(suite, version, arch string, components []str
...
-		Label:   "Ubuntu",
-		PrivKey: s.privKey,
+		Suite:       suite,
+		Version:     version,
+		Label:       "Ubuntu",
+		PrivKey:     s.privKey,
+		DigestKinds: []string{"SHA256"},
+	}
+	if adjustRelease != nil {
+		adjustRelease(release)
```
Comment (upils) [inline internal/archive/archive_test.go]: [Note to reviewer]: This was moved before preparing the packages so they pick up the adjusted DigestKinds.

File: internal/archive/testarchive/testarchive.go (hunk 19 above)
Comment (upils) [reply internal/archive/testarchive/testarchive.go]: I changed it back to setting a slice of DigestKinds in `Package` to avoid all these drawbacks. This is consistent with how we deal with other fields (especially `arch`) . Only `prepareArchiveAdjustRelease` considers that both `Release.DigestKinds` and `Package.DigestKinds` should be in sync by default, but callers are free to tweak that If not using `prepareArchiveAdjustRelease`, callers are also responsible to properly fill `Release` and `Package`. I think this is a better separation of concerns and is more consistent with the existing code.

File: internal/archive/testarchive/testarchive.go (hunk 21 above)
Comment (upils) [reply internal/archive/testarchive/testarchive.go]: @FinnRG answer to these questions: https://github.com/canonical/chisel/pull/306#discussion_r3570413066. I do not really trust GH linking so I am pasting the answer here:

> The ftpmaster publisher always rebuilds the entire Release (and associated index files) all at once, so there was no transition phase.

> Just to be clear, the hashes inside the Packages and Sources index files may be different from the hash(es) used in the Release file and therefore the by-hash lookup.

As the used hash is a release-level decision, I am moving the call `strongestKind` before the call to `Walk` to make it obvious.

**Comment by lczyk** [comment]
i agree with @upils changes 👍 have reached roughly the same conclusion / direction when looking at this

File: internal/archive/testarchive/testarchive.go (hunk 20 above)
Comment (upils) [reply internal/archive/testarchive/testarchive.go]: The aim here is to be consistent with the promise made by a real valid archive: when supporting by-hash URL, publish them at least for the strongest supported hash in the release. This piece of code is unrelated to how the client should behave when consuming the archive.

Regarding the actual behavior of Chisel: when trying to fetch the package index, `findDigest` is called to get the strongest hash associated to the file, and this hash is then used to try to get the file "by-hash". In the scenario where the archive published multiple hashes for files, Chisel only tries to use the "by-hash" URL with the strongest hash it supports and falls back to the regular URLs if encounters a 404 error. This is consistent with apt's behavior.

**Comment by lczyk** [comment]
marking as `priority` since its blocking all of chisel-releases 26.10 work

File: internal/archive/archive_test.go (hunk 22 above)
Comment (niemeyer) [reply internal/archive/archive_test.go]: This feels odd. Why would a function A pass a function B into a function C, when all C is doing is calling B with A's data?  The whole point of having this as a parameter, in my limited understanding, is to have the _final_ release be adjusted, with the data that A does _not know_ about.

Can you please sync with Alberto on this?  This is a blocker because these tests are not simply unit tests, AIUI - some of the tests in this package are the tests that are actually exercised as well on the real archives to make sure that they remain sane, in our GitHub workflows. I'm not sure if this is touching them or not, but needs verifying as well.

File: internal/archive/archive_test.go [hunk 23]
```diff
@@ -155,10 +162,10 @@ func (s *httpSuite) prepareArchiveAdjustRelease(suite, version, arch string, com
 	if err != nil {
 		panic(err)
 	}
-	if adjustRelease != nil {
-		adjustRelease(release)
+	err = release.Render(base.Path, s.responses)
+	if err != nil {
+		panic(err)
```
Comment (niemeyer) [inline internal/archive/archive_test.go:167]: This wasn't here before. It's probably a good thing to add it, but please just confirm that this is okay given the context I just mentioned.

**Review by niemeyer — CHANGES_REQUESTED** [review]
Thanks for the changes made. This is much nicer and almost ready. Just a couple of details:

File: internal/archive/archive_test.go (hunk 22 above)
Comment (upils) [reply internal/archive/archive_test.go]: > This feels odd. Why would a function A pass a function B into a function C, when all C is doing is calling B with A's data? The whole point of having this as a parameter, in my limited understanding, is to have the _final_ release be adjusted, with the data that A does _not know_ about.

Yes conceptually it made little sense. I reworked it, putting back the call to `adjustRelease` where it was and passing `digestKinds` as an argument to `prepareArchiveAdjustRelease`. 

> Can you please sync with Alberto on this? This is a blocker because these tests are not simply unit tests, AIUI - some of the tests in this package are the tests that are actually exercised as well on the real archives to make sure that they remain sane, in our GitHub workflows. I'm not sure if this is touching them or not, but needs verifying as well.

We discussed the whole issue with Alberto. `prepareArchiveAdjustRelease` and `prepareArchive` are not used in real archive tests, so these changes do not affect them.

File: internal/archive/archive_test.go (hunk 23 above)
Comment (upils) [reply internal/archive/archive_test.go:167]: As mentioned in the other comment https://github.com/canonical/chisel/pull/306#discussion_r3821974654 this should not affect real archive test so this should be okay and should not risk hiding failures.

File: internal/archive/testarchive/testarchive.go [hunk 24]
```diff
@@ -212,8 +234,28 @@ func (pi *PackageIndex) Content() []byte {
...
 }
 
-func makeSha256(b []byte) string {
-	return fmt.Sprintf("%x", sha256.Sum256(b))
+// digestKinds lists the digest kinds this package can render, strongest first.
+var digestKinds = []string{"SHA512", "SHA256"}
+
+// strongestKind returns the strongest of kinds, or "" if kinds holds none of
+// digestKinds.
+func strongestKind(kinds []string) string {
```
Comment (niemeyer) [inline internal/archive/testarchive/testarchive.go:242]: This is error prone as we are passing a list of kinds to a function asking which one is strongest and then potentially getting an empty string back. Very easy to make mistakes.

File: internal/archive/testarchive/testarchive.go [hunk 25]
```diff
@@ -212,8 +234,28 @@ func (pi *PackageIndex) Content() []byte {
...
+}
+
+func makeDigest(kind string, b []byte) string {
+	switch kind {
+	case "SHA512":
+		return fmt.Sprintf("%x", sha512.Sum512(b))
+	case "SHA256":
+		return fmt.Sprintf("%x", sha256.Sum256(b))
+	}
+	panic("unknown digest kind: " + kind)
```
Comment (niemeyer) [inline internal/archive/testarchive/testarchive.go:258]: This is fine because it is a test, otherwise the issue would be similar.

File: internal/archive/archive.go [hunk 26]
```diff
@@ -328,10 +331,47 @@ func (index *ubuntuIndex) fetchRelease() error {
...
+
+// findDigest returns the digest recorded for path in the release, along with
+// the field it was found in, trying the given fields in order.
+func findDigest(release control.Section, path string, order []digestField) (digest string, field digestField) {
+	for _, f := range order {
+		if d, _, ok := control.ParsePathInfo(release.Get(f.name), path); ok {
+			return d, f
+		}
+	}
+	return "", digestField{}
```
Comment (niemeyer) [inline internal/archive/archive.go:358]: Let's please have an explicit (ok bool) as the last result, so call sites are "forced" to deal with the missing cases.

File: internal/archive/archive.go [hunk 27]
```diff
@@ -328,10 +331,47 @@ func (index *ubuntuIndex) fetchRelease() error {
...
+
+func packageDigest(section control.Section) (digest string, kind cache.DigestKind) {
+	for _, f := range digestFields {
+		if d := section.Get(f.name); d != "" {
+			return d, f.kind
+		}
+	}
+	// No digest advertised; fall back to SHA256 so the package can still be
+	// cached and retrieved by its computed digest.
+	return "", cache.SHA256
```
Comment (niemeyer) [inline internal/archive/archive.go:369]: Same.. and this is even more error prone. We're returning a partially invalid value and expecting every call site to deal with the fact of the half valid information, knowing which one is valid and how.

**Review by niemeyer — APPROVED** [review]
Thanks for these fixes, and sorry to be slow here. This got caught in the middle of company holidays and events, and got more delayed than it should have been.

There are still open details to be addressed in this PR, @upils, but I'm going ahead and merging it as-is as those are nitpicks/polishings, and this is blocking the release.
